tests/state/ViewerState/backdrop_slice.test.ts

```diff
@@ -0,0 +1,123 @@
+      const { result } = renderHook(() => useViewerStateStore());
+
+      // Initialized as null
+      expect(result.current.backdropKey).toBeNull();
+
+      act(() => {
+        result.current.setBackdropKey(MOCK_DATASET_WITHOUT_BACKDROP, "test");
+      });
+      expect(result.current.backdropKey).toBeNull();
+    });
+
+    it("allows setting backdrop keys that are in the dataset.", () => {
+      const mockDataset = {
+        hasBackdrop: (key: string) => key === "test1" || key === "test2",
+        getDefaultBackdropKey: () => "test1",
+      } as unknown as Dataset;
+      const { result } = renderHook(() => useViewerStateStore());
+
+      // Should initialize to default backdrop key
+      act(() => {
+        result.current.setBackdropKey(mockDataset, "test1");
+      });
+      expect(result.current.backdropKey).toBe("test1");
+
+      // Can set another valid backdrop key
+      act(() => {
+        result.current.setBackdropKey(mockDataset, "test2");
+      });
+      expect(result.current.backdropKey).toBe("test2");
+
+      // Ignores keys that do not exist
+      act(() => {
+        result.current.setBackdropKey(mockDataset, "test3");
+      });
+      expect(result.current.backdropKey).toBe("test2");
+    });
+  });
+
+  describe("setBackdropVisible", () => {
+    it("does not allow visibility to be enabled when backdrop key is null", () => {
+      const { result } = renderHook(() => useViewerStateStore());
+
+      // Does not allow visibility to be enabled when backdrop key is null
+      act(() => {
+        result.current.setBackdropVisible(true);
+      });
+      expect(result.current.backdropKey).toBeNull();
+      expect(result.current.backdropVisible).toBe(false);
+    });
+
+    it("toggles visibility", () => {
+      // Directly set state
+      const { result } = renderHook(() => useViewerStateStore());
+      act(() => {
+        useViewerStateStore.setState({ backdropKey: "test" });
+      });
+
+      act(() => {
+        result.current.setBackdropVisible(true);
+      });
+      expect(result.current.backdropVisible).toBe(true);
+      act(() => {
+        result.current.setBackdropVisible(false);
+      });
+      expect(result.current.backdropVisible).toBe(false);
+    });
+  });
+
+  it("can set backdrop properties with clamping", () => {
+    const { result } = renderHook(() => useViewerStateStore());
+    act(() => {
+      result.current.setBackdropBrightness(75);
+      result.current.setBackdropSaturation(50);
+      result.current.setObjectOpacity(25);
+    });
+    expect(result.current.backdropBrightness).toBe(75);
+    expect(result.current.backdropSaturation).toBe(50);
+    expect(result.current.objectOpacity).toBe(25);
+
+    act(() => {
+      result.current.setBackdropBrightness(220);
+      result.current.setBackdropSaturation(150);
+      result.current.setObjectOpacity(120);
+    });
+    expect(result.current.backdropBrightness).toBe(200);
```

Comment:
Nit: these min and max values should probably be constants somewhere, maybe?   This is part of that data validation thing we discussed where I'm not sure the actions should be validating the inputs.  I wonder if there is a industry consensus on this.

 I suppose maybe clamping is better than not clamping.  But then below I see you throw for NaNs, but clamp for out of range values?

Reply:
I read some discussions on this! The general sense I got was that you should be validating inputs at multiple levels. It's good to do it from the UI so that users can get immediate feedback, but bad data passed in your state model usually indicates a developer error and should throw errors/warnings to help you catch it.

Agree on making constants for the min/max! I could probably make out-of-range values invalid as well.